### ISSUE TITLE
3.0 - Trashed orders cannot be permanently deleted with bulk actions

### ISSUE THREAD

**Issue by robincornett** [easydigitaldownloads/easy-digital-downloads#7937]
### Expected behavior
On the orders list table, I expect to have the same basic bulk actions available to me that I do in other list tables (such as posts).

### Actual behavior
When viewing "Trashed" orders, there is no way to delete all of them. Orders (applies to refunds as well):

![image](https://user-images.githubusercontent.com/4853066/82370101-fed2a680-99e5-11ea-9f7e-a06744417131.png)

Viewing "Trashed" downloads:

![image](https://user-images.githubusercontent.com/4853066/82370228-36415300-99e6-11ea-8ef9-d96df5c14778.png)

(I can Delete Permanently by selecting multiple items, or click the "Empty Bin/Trash" button and clean everything out in one go.

### Steps to reproduce the behavior
1) In EDD 3.0, delete some orders. This now moves them to a temporary "trash" location, added in 3.0.
2) Now go to the trashed orders (`/wp-admin/edit.php?post_type=download&page=edd-payment-history&status=trash`)
3) The trashed orders cannot be deleted in bulk (the action links are available, so individual orders can be deleted permanently, although the redirect takes me out of the trashed orders view).

### Information (if a specific version is affected):
PHP Version: 7.3.5
EDD Version (or branch): 3.0
WordPress Version: 5.4

**Comment by ashleyfae** [comment]
Wasn't there an old Slack discussion about whether or not we _should_ allow orders to be truly deleted? I don't remember what the consensus was but I recall this discussion happening...

@cklosowski @JJJ (can't remember if you were in on it)

**Comment by JJJ** [comment]
This is a complicated issue, both in terms of discussion and code. I'll do my best to add context.

## TL;DR
At the time, we simply could not get all of the combinations working reliably. Also, WordPress does not work this way currently anywhere; you cannot Untrash to any status - you can only "Restore" or "Delete Permanently".

For example:
* Create a Post
* Change the status to "Private"
* Publish it
* Trash it
* View Trash
* Notice that "Restore" and "Delete Permanently" are the only options
<img width="242" alt="jjj-sc-2020-05-26 at 08 25 02@2x" src="https://user-images.githubusercontent.com/88951/82906145-6cbd2780-9f2a-11ea-9b70-f6696960c4a7.png">

----

## GAAP
Much of the of the way that Orders works is designed to follow [Generally Accepted Accounting Principles](https://en.wikipedia.org/wiki/Generally_Accepted_Accounting_Principles_(United_States)), ironically without following them explicitly when it comes to things like function names and code references, largely for backwards compatibility and to keep things relatively easy to follow compared to needing everyone to become an CPA.

In the world of accounting, there are some transactions that just do not make sense. (For example, revoking a refund.) While WordPress does have a loose API for transitioning between statuses, it requires us to manually address what happens when transitioning from every status to every other status, including Trash.

To simplify this process for ourselves, in an attempt to reduce our own code errors with transitioning between an array of exponentially growing statuses, we followed the WordPress/bbPress route of storing the original status in meta, and only allowing untrashing back to its original status.

----

## Permanence

Regarding deleting anything permanently, it goes against one of the more important accounting principles of permanence, but is also a very WordPress thing to allow to happen. After several exciting discussions in Slack, the decision was ultimately made to allow users to permanently delete records.

## Delete vs. Destroy

Internally, this introduced an API naming problem, as anything with `_delete_` is reserved for single database rows, and not intended to cascade down and throughout child rows, or relative rows in other database tables, so `_destroy_` functions were introduced to obliterate all dependent records.

`edd_destroy_customer()` is a good example of this. To avoid orphaning email addresses and physical addresses, this function exists to also permanently delete those relative database objects.

Note, however, that destroying a customer still _does not_ delete their Orders.

At this point in 3.0 development, I am unsure if that is by design or an oversight. During my time on 3.0, it was by design, because Orders were permanent then and could never be deleted. (Orders were part of the ledger, and corrections to the ledger were just reversals.)

Now that Orders can be permanently deleted, should they be deleted when the customer that made them is permanently deleted? Are we confident that we've addressed every UI element that expects a Customer to exist?
----

## Conclusion
Back to this issue specifically, my recommendation is that implementing this is both complicated and error prone. Any-to-any status changes should come with a smorgasbord of unit tests to verify that all of the accounting is accurate.

**Comment by JJJ** [comment]
I think this is a two-part issue.

1. is "Delete Permanently" a bulk action we want to allow?
1. should items in the Trash be allowed to be given any status when transitioning out of Trash?

Admittedly, I have a high degree of anxiety about both.

## TL;DR
At the time, regarding 1 above, the thought was that obliterating everything in the trash was potentially irreversibly disastrous. We kept "intentional friction" to try and invisibly hint to a shop owner that permanently deleting records isn't something that should happen very often, if ever, and that errors should be corrected instead of being wiped away.

Regarding number 2 above, there were 2 reasons:

* we could not get all of the status transitions working reliably
* WordPress does not work this way currently anywhere; you cannot Untrash to any status - you can only "Restore" or "Delete Permanently"

For example:
* Create a Post
* Change the status to "Private"
* Publish it
* Trash it
* View Trash
* Notice that "Restore" and "Delete Permanently" are the only options
<img width="242" alt="jjj-sc-2020-05-26 at 08 25 02@2x" src="https://user-images.githubusercontent.com/88951/82906145-6cbd2780-9f2a-11ea-9b70-f6696960c4a7.png">

----

## GAAP
Much of the of the way that Orders works is designed to follow [Generally Accepted Accounting Principles](https://en.wikipedia.org/wiki/Generally_Accepted_Accounting_Principles_(United_States)), ironically without following them explicitly when it comes to things like function names and code references, largely for backwards compatibility and to keep things relatively easy to follow compared to needing everyone to become an CPA.

In the world of accounting, there are some transactions that just do not make sense. (For example, revoking a refund.) While WordPress does have a loose API for transitioning between statuses, it requires us to manually address what happens when transitioning from every status to every other status, including Trash.

To simplify this process for ourselves, in an attempt to reduce our own code errors with transitioning between an array of exponentially growing statuses, we followed the WordPress/bbPress route of storing the original status in meta, and only allowing untrashing back to its original status.

----

## Permanence

Regarding deleting anything permanently, it goes against one of the more important accounting principles of permanence, but is also a very WordPress thing to allow to happen. After several exciting discussions in Slack, the decision was ultimately made to allow users to permanently delete records.

## Delete vs. Destroy

Internally, this introduced an API naming problem, as anything with `_delete_` is reserved for single database rows, and not intended to cascade down and throughout child rows, or relative rows in other database tables, so `_destroy_` functions were introduced to obliterate all dependent records.

`edd_destroy_customer()` is a good example of this. To avoid orphaning email addresses and physical addresses, this function exists to also permanently delete those relative database objects.

Note, however, that destroying a customer still _does not_ delete their Orders.

At this point in 3.0 development, I am unsure if that is by design or an oversight. During my time on 3.0, it was by design, because Orders were permanent then and could never be deleted. (Orders were part of the ledger, and corrections to the ledger were just reversals.)

Now that Orders can be permanently deleted, should they be deleted when the customer that made them is permanently deleted? Are we confident that we've addressed every UI element that expects a Customer to exist?

----

## Conclusion
Back to this issue specifically, bulk permanent deletion seems fine to implement so long as all of the related accounting can be double checked for correctness, likely with unit tests to confirm.

I think transitioning out of Trash to any status is going to seem easy to implement on the UI side, but very challenging on the accounting side. It's also not how WordPress works, so it may unintentionally introduce more confusion if EDD goes above and beyond to allow for it.

## Epilogue
I blame [WordPress for all of this.](https://core.trac.wordpress.org/ticket/12706) The lack of a common API and shared `Status` class prevents every application inside of it from having any obvious and logical path to follow when said status changes. Everything needing to be hard-coded makes transitioning between statuses a manual and error prone process of elimination.

We should be able to "Trash" a Customer. And we should be able to have a `EDD\Customer\Status\Trash` class that knows how to react when a Customer goes in or out of `trash` status. We could more easily have a "Suspend" status then, etc... But a lack of any `Status` type class makes it impossible.

**Comment by cklosowski** [comment]
All this above is accurate @JJJ and along with it we determined that a bulk delete in the UI is a dangerous approach to the sales records, and that if you are bulk deleting orders you are better off using the EDD Recalculation tools which contains a 'Reset Store' which not only destroys the sales, but all records except for your products, essentially letting someone start back at zero.

**Comment by cklosowski** [comment]
@easydigitaldownloads/developers I would like to re-open this and have more of a discussion about if we should revert this change.

**Comment by cklosowski** [comment]
My vote, is that we revert this change and not support bulk deleting of orders in the trash. The most commonly used case for this is people removing test data, and as such, we offer a tool for that which is far more efficient, and touches on areas of the database that this does not. For instance, allowing this method has the chance to cause a timeout due to the amount of data being deleted, and could leave orphaned rows. The reset tool has built in batching to allow us to not run into those types of errors.

**Comment by spencerfinnell** [comment]
I get the idea of the "forced friction" but I think @robincornett finding the Bulk Action item missing and my quickness to fix what _seemed to be_ a small bug potentially shows that the forced friction might result in support tickets/confusion when users find they can't perform an action -- that is allowed normally -- in bulk.

Perhaps similar to the messaging and icons we added when viewing an Order that it cannot be directly edited we add a note explicitly explaining why we have opted to disallow bulk deletions.

Maybe a confirmation dialog when clicking a single "Delete" that explains that deleting is bad, but if you need to do it to remove test Data, do it via the reset tools.

**Comment by robincornett** [comment]
If we are removing the bulk delete for orders, should that action be removed from individual trashed orders as well?

Would it be possible to leave the "Delete Permanently" option in the list, but instead of destroying all of the orders, as the PR does, redirect the user to the reset tools (I did not know this existed...but would the reset tools clear out the entire store, or just trashed orders?). Cleaning out the trash is what I was originally trying to do when I raised this issue.

**Comment by cklosowski** [comment]
I think a store owner should have the ability to delete individual orders (as they could before). there is a time and place for it, albeit, not ideal.

We have a problem in that we've allowed that for a long time and I think that people would revolt if we fully removed it.

I don't really want to 'redirect' people there, but we could up our documentation to have a document entitled 'bulk deleting orders' that tells people about the reset tool, and gives some insight on why we don't support bulk delete from the UI (which is to help protect them from accidentally removing records of truth.

**Comment by spencerfinnell** [comment]
> We have a problem in that we've allowed that for a long time and I think that people would revolt if we fully removed it.

> I don't really want to 'redirect' people there

@cklosowski I think an "Are you sure you want to do this?" dialog that can link out to more documentation or allow a continuation of destruction is a good compromise. 

Looking at core I'm surprised there isn't an AYS for fully deleting a normal Post/Page and it seems even more important for communicating the outcome for Orders.

We essentially have one in place for deleting a Customer. Clicking "Delete" on the list table takes you to another page which requires two additional interactions before it can be deleted.

<img width="654" alt="Screen Shot 2020-05-26 at 2 23 17 PM" src="https://user-images.githubusercontent.com/491124/82936372-7c556400-9f5c-11ea-8be6-f5964d1c0b20.png">

**Comment by cklosowski** [comment]
I'd love to get an opinion from @ashleyfae. This is a pretty big decision (even though it may not seem like it) as we have the opportunity to prevent what our support sees from time to time where people will erroneously delete things and not be able to recover.

I'd even go so far as to build out the UI around the caps we've built in so that the delete order doesn't even show up if they aren't a store manager. Much like needing the managers code at a register to perform certain actions.

But I'ld like to wait for Ashely's opinion on the matter before we move forward.

**Comment by ashleyfae** [comment]
I think we should revert the change. I'd rather not have it in and reevaluate the decision later if necessary, than add it in and regret it. This is an easy addition to the docs, and easy to add a saved reply in support. It's also very easy to add back in later if we change our minds for whatever reason. But for now I'd rather be overly cautious.

**Comment by robincornett** [comment]
I'm not sure how a reversion works, but I would suggest that a couple of the changes made not be reverted:
- changing the "Trash" bulk action label to be "Move to Trash" (for parity with WordPress Core)
- not attempting pagination if no orders match a view

Question: should bulk actions within trashed orders revert back to the original list, then, which was to bulk set multiple orders to a specific status (Mark Completed, Mark Pending, etc.)?

**Comment by cklosowski** [comment]
@robincornett I'm good with those two items to _not_ revert. I think that makes sense.

As far as the question:
Yes we should still allow the 'All' list to set items as complete or pending in a bulk manor.

**Comment by robincornett** [comment]
@cklosowski one of the changes was also to allow trashed orders to be permanently deleted on an individual basis, which was not present in 3.0 before this. I've reverted the bulk actions, but should I also revert the individual delete action, or leave that in? (So individual actions would be "Restore" or "Delete Permanently" instead of just "Restore".)

Note: reverting the bulk actions to the full list means that more bulk actions are available than individual actions--because I can mark something as Pending, Complete, Preapproved, etc. from bulk actions, but not from the row's action links. Not sure if that matters, but it's there.

**Comment by cklosowski** [comment]
Regarding the more bulk actions, that is fine. It is how 2.9 is done as well, we don't add all the status options to the row actions.

As for the `Delete Permanently`, I _think_ I can be ok with that, but I think that decision should include @ashleyfae's opinion as well.

**Comment by ashleyfae** [comment]
I'm okay with keeping it.

**Comment by ashleyfae** [comment]
I agree with the above, and with this:

> I wonder if the trashed orders bulk actions should be limited to simply "Restore" instead of the standard list of actions, since restoring the order will check the order meta and automatically change the restored order to its pending/cancelled/complete/etc. state, instead of allowing the bulk actions to potentially significantly change that.

**Comment by WildeWebMarketing** [comment]
I found this thread after realizing I have to manually delete trashed orders that were initially Abandoned. The reason I want to delete the Abandoned, which are now in Trash? Because on the customer's [download_history] page, it shows the product names and under files, it says "Status: Abandoned" or "Status: Trash" and that just looks stupid.

We don't need excessive hand-holding. You removed bulk actions because some store owner might not know what they are doing and delete orders by accident? It's our store, we shouldn't be barred from running it as we see fit. That's like Microsoft Windows removing the ability to delete stuff because my mom isn't good with computers.

**Comment by Digit-89** [comment]
Hello,
I agree with the above statement and in addition:
It makes no sense to force website admins to store forever abandoned, failed, pending and trashed orders. There should be a tool to bulk delete them with a few filters like "Keep the orders from the last 1-3-6-12 months".
I am running EDD for 6 months only and have 100K+ orders, I need to clean what's unnecessary.

**Comment by burninc0de** [comment]
Couldn't agree more with the above two statements. I understand the reasoning for disabling bulk delete by default, but at least give us an option to enable it somewhere in the options. Because the only alternatives at this point are manual removal (do admins really need this headache?) or PhPMyAdmin (far "dangerous" imho than just giving us the bulk delete back)
